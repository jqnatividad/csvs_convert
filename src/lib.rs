use snafu::{Snafu, ensure};
use snafu::prelude::*;
use serde_json::Value;
use std::fs::File;
use std::path::PathBuf;
use std::io::BufReader;
use csv::Writer;

#[non_exhaustive]
#[derive(Debug, Snafu)]
pub enum Error {
    #[snafu(display("{}", message))]
    DatapackageMergeError { message: String },

    #[snafu(display("Error reading file {}: {}", filename, source))]
    IoError { source: std::io::Error, filename: String },

    #[snafu(display("Error parsing JSON {}: {}", filename, source))]
    JSONError { source: serde_json::Error, filename: String },

    #[snafu(display("Error loading zip file {}: {}", filename, source))]
    ZipError { source: zip::result::ZipError, filename: String },

    #[snafu(display("Error loading zip file {}: {}", filename, source))]
    CSVError { source: csv::Error, filename: String }
}

fn make_mergeable_resource(mut resource: Value) -> Result<Value, Error>{
    let mut fields = resource["schema"]["fields"].take();
    let fields_option = fields.as_array_mut();

    ensure!(fields_option.is_some(), DatapackageMergeSnafu {message: "Datapackages need a `fields` list"});

    let mut new_fields = serde_json::Map::new();
    for field in fields_option.unwrap().drain(..) {
        let name_option = field["name"].as_str();
        ensure!(name_option.is_some(), DatapackageMergeSnafu {message: "Each field needs a name"});
        new_fields.insert(name_option.unwrap().to_owned(), field);
    };

    resource["schema"].as_object_mut().unwrap().insert("fields".to_string(), new_fields.into());

    Ok(resource)
}

fn make_mergeable_datapackage(mut value: Value) -> Result<Value, Error> {

    let mut resources = value["resources"].take();

    let resources_option = resources.as_array_mut();
    ensure!(resources_option.is_some(), DatapackageMergeSnafu {message: "Datapackages need a `resources` key as an array"});

    let mut new_resources = serde_json::Map::new();
    for resource in resources_option.unwrap().drain(..) {
        let path;
        {
            let path_str = resource["path"].as_str();
            ensure!(path_str.is_some(), DatapackageMergeSnafu{message: "datapackage resource needs a name or path"});
            path = path_str.unwrap().to_owned();
        }

        let new_resource = make_mergeable_resource(resource)?;
        new_resources.insert(path, new_resource);
    };

    value.as_object_mut().unwrap().insert("resources".into(), new_resources.into());

    Ok(value)
}


fn make_datapackage_from_mergeable(mut value: Value) -> Result<Value, Error> {

    let mut resources = value["resources"].take();

    let resources_option = resources.as_object_mut();

    let mut new_resources = vec![];
    for resource in resources_option.unwrap().values_mut() {
        let new_resource = make_resource_from_mergable(resource.clone())?;
        new_resources.push(new_resource);
    };

    value.as_object_mut().unwrap().insert("resources".into(), new_resources.into());

    Ok(value)
}

fn make_resource_from_mergable(mut resource: Value) -> Result<Value, Error>{
    let mut fields = resource["schema"]["fields"].take();
    let fields_option = fields.as_object_mut();

    let mut new_fields = vec![];
    for field in fields_option.unwrap().values_mut() {
        new_fields.push(field.clone());
    };

    resource["schema"].as_object_mut().unwrap().insert("fields".to_string(), new_fields.into());

    Ok(resource)
}


fn datapackage_json_to_value (filename: &str) -> Result<Value, Error> {
    if filename.ends_with(".json") {
        let file = File::open(&filename).context(IoSnafu {filename})?;
        let json: Value = serde_json::from_reader(BufReader::new(file)).context(JSONSnafu{filename})?;
        Ok(json)
    } else if filename.ends_with(".zip") {
        let file = File::open(&filename).context(IoSnafu {filename})?;
        let mut zip = zip::ZipArchive::new(file).context(ZipSnafu {filename})?;
        let zipped_file = zip.by_name("datapackage.json").context(ZipSnafu {filename})?;
        let json: Value = serde_json::from_reader(BufReader::new(zipped_file)).context(JSONSnafu{filename})?;
        Ok(json)
    } else if PathBuf::from(filename).is_dir() {
        let mut path = PathBuf::from(filename);
        path.push("datapackage.json");
        let file = File::open(path).context(IoSnafu {filename})?;
        let json: Value = serde_json::from_reader(BufReader::new(file)).context(JSONSnafu{filename})?;
        Ok(json)
    } else {
        Err(Error::DatapackageMergeError {message: "could not detect a datapackage".into()})
    }
}

fn merge_datapackage_json(mut base: Value, mut merger: Value) -> Result<Value, Error> {
    let merger_resources_value = merger["resources"].take();

    let merger_resources = merger_resources_value.as_object().unwrap();
    let base_resources = base["resources"].as_object_mut().unwrap();

    for (resource, resource_value) in merger_resources {
        if !base_resources.contains_key(resource) {
            base_resources.insert(resource.clone(), resource_value.clone());
        } else {
            for (field, field_value) in resource_value["schema"]["fields"].as_object().unwrap() {
                ensure!(field_value.is_object(), DatapackageMergeSnafu {message: "Each field needs to be an object"});

                let base_fields = base_resources[resource]["schema"]["fields"].as_object_mut().unwrap();

                if !base_fields.contains_key(field) {
                    base_fields.insert(field.clone(), field_value.clone());
                } else {
                    ensure!(base_fields[field].is_object(), DatapackageMergeSnafu {message: "Each field needs to be an object"});
                    let base_fieldinfo = base_fields[field].as_object_mut().unwrap();

                    let base_type = base_fieldinfo["type"].as_str().unwrap_or_default();
                    let field_type = field_value["type"].as_str().unwrap_or_default();

                    if field_type != base_type || base_type.is_empty() || field_type.is_empty() {
                        base_fieldinfo.insert("type".to_string(), "string".into());
                    }

                    let base_count = base_fieldinfo["count"].as_u64().unwrap_or_default();
                    let field_count = field_value["count"].as_u64().unwrap_or_default();
                    
                    if base_count > 0 && field_count > 0 {
                        base_fieldinfo.insert("count".to_string(), (field_count + base_count).into());
                    }
                }
            }
        }
    }
    Ok(base)
}


pub fn merge_datapackage_jsons(datapackages: Vec<String>) -> Result<Value, Error> {
    ensure!(datapackages.len() > 1, DatapackageMergeSnafu {message: "Need more 2 or more datapackages"});
    let mut merged_value = make_mergeable_datapackage(datapackage_json_to_value(&datapackages[0])?)?;

    for file in datapackages[1..].iter() {
        merged_value = merge_datapackage_json(merged_value, make_mergeable_datapackage(datapackage_json_to_value(file)?)?)?;
    }

    make_datapackage_from_mergeable(merged_value)
}


pub fn write_merged_csv(
    csv_reader: csv::Reader<impl std::io::Read>, 
    mut csv_writer: Writer<File>,
    resource_fields: &std::collections::HashMap<String, usize>,
    output_fields: &[String]
) -> Result<Writer<File>, Error> {
    let output_map: Vec<Option<&usize>> = output_fields.iter().map(|field| {resource_fields.get(field)}).collect();
    let output_map_len = output_map.len();
    for row in csv_reader.into_records() {
        let mut output_row = Vec::with_capacity(output_map_len);
        let row = row.unwrap();
        for item in &output_map {
            match item {
                Some(index) => {output_row.push(row.get(**index).unwrap())},
                None => {output_row.push("")}
            }
        }
        csv_writer.write_record(output_row).unwrap();
    }
    Ok(csv_writer)
}


pub fn merge_datapackage(datapackages: Vec<String>, output_path: PathBuf) -> Result<(), Error> {
    ensure!(datapackages.len() > 1, DatapackageMergeSnafu {message: "Need more 2 or more files"});
    
    std::fs::create_dir_all(&output_path).context(IoSnafu {filename: output_path.to_string_lossy()})?;

    let mut merged_datapackage_json = merge_datapackage_jsons(datapackages.clone())?;

    let path = PathBuf::from(&output_path);

    let datapackage_json_path_buf = path.join("datapackage.json");

    let writer = File::create(&datapackage_json_path_buf).context(IoSnafu {filename: datapackage_json_path_buf.to_string_lossy()})?;

    serde_json::to_writer_pretty(writer, &merged_datapackage_json).context(JSONSnafu {filename: datapackage_json_path_buf.to_string_lossy()})?;

    let mut csv_outputs = std::collections::HashMap::new();
    let mut output_fields = std::collections::HashMap::new();

    for resource in merged_datapackage_json["resources"].as_array_mut().unwrap() {

        let mut field_order_map = serde_json::Map::new();
        let mut fields: Vec<String> = Vec::new();
        for (index, field) in resource["schema"]["fields"].as_array().unwrap().iter().enumerate() {
            let name = field["name"].as_str().unwrap();
            field_order_map.insert(name.into(), index.into());
            fields.push(name.to_owned());
        }

        let resource_path = resource["path"].as_str().unwrap().to_owned();

        let mut full_path = path.join(&resource_path);
        full_path.pop();
        std::fs::create_dir_all(&full_path).context(IoSnafu {filename: full_path.to_string_lossy()})?;

        let mut writer = Writer::from_path(path.join(&resource_path)).context(CSVSnafu {filename: &resource_path})?;
        writer.write_record(fields.clone()).context(CSVSnafu {filename: &resource_path})?;
        csv_outputs.insert(resource_path.clone(), writer);

        output_fields.insert(resource_path.clone(), fields);

        resource.as_object_mut().unwrap().insert("field_order_map".into(), field_order_map.into());
    }

    for file in datapackages.iter() {
        let mut datapackage_json = datapackage_json_to_value(file)?;
        for resource in datapackage_json["resources"].as_array_mut().unwrap() {
            let mut resource_fields = std::collections::HashMap::new();
            for (num, field) in resource["schema"]["fields"].as_array().unwrap().iter().enumerate() {
                resource_fields.insert(field["name"].as_str().unwrap().to_owned(), num);
            }
            let resource_path = resource["path"].as_str().unwrap().to_owned();

            let mut csv_output = csv_outputs.remove(&resource_path).unwrap();

            let output_fields = output_fields.get_mut(&resource_path).unwrap();

            if file.ends_with(".json") {
                let mut file_pathbuf = PathBuf::from(file);
                file_pathbuf.pop();
                file_pathbuf.push(&resource_path);
                let csv_reader = csv::Reader::from_path(file_pathbuf).unwrap();
                csv_output = write_merged_csv(csv_reader, csv_output, &resource_fields, output_fields).unwrap();
            } else if file.ends_with(".zip") {
                let zip_file = File::open(&file).context(IoSnafu {filename: file})?;
                let mut zip = zip::ZipArchive::new(zip_file).context(ZipSnafu {filename: file})?;
                let zipped_file = zip.by_name(&resource_path).context(ZipSnafu {filename: file})?;
                let csv_reader = csv::Reader::from_reader(zipped_file);
                csv_output = write_merged_csv(csv_reader, csv_output, &resource_fields, output_fields).unwrap();
            } else if PathBuf::from(file).is_dir() {
                let path = PathBuf::from(file);
                let path = path.join(&resource_path);
                let csv_reader = csv::Reader::from_reader(File::open(path).unwrap());
                csv_output = write_merged_csv(csv_reader, csv_output, &resource_fields, output_fields).unwrap();
            } else {
                return Err(Error::DatapackageMergeError {message: "could not detect a datapackage".into()})
            }

            csv_outputs.insert(resource_path, csv_output);
        }
    }

    Ok(())
}


#[cfg(test)]
mod tests {
    use super::*;
    use insta;
    use tempfile::TempDir;
    use std::io::BufRead;

    fn test_merged_csv_output(tmp: &PathBuf, name: String) {
        let csv_dir = tmp.join("csv");
        let paths = std::fs::read_dir(csv_dir).unwrap();
        for path in paths {
            let path = path.unwrap().path();
            let file_name = path.file_name().unwrap().to_string_lossy().into_owned();
            let test_name = format!("{name}_{file_name}");
            let file = File::open(path).unwrap();
            let lines: Vec<String> = std::io::BufReader::new(file).lines().map(|x| x.unwrap()).collect();
            insta::assert_yaml_snapshot!(test_name, lines);
        }

    }

    fn test_datapackage_merge(name: &str, datapackage1: &str, datapackage2: &str) {
        {
            let tmp_dir = TempDir::new().unwrap();
            let tmp = tmp_dir.path().to_owned();

            merge_datapackage(
                    vec![format!("fixtures/{datapackage1}/datapackage.json"), format!("fixtures/{datapackage2}/datapackage.json")],
                    tmp.clone()
            ).unwrap();

            insta::assert_yaml_snapshot!(format!("{name}_json"),
                datapackage_json_to_value(&tmp.to_string_lossy().into_owned()).unwrap()
            );
            test_merged_csv_output(&tmp, format!("{name}_json"))
        }

        {
            let temp_dir = TempDir::new().unwrap();
            let tmp = temp_dir.path().to_path_buf();

            merge_datapackage(
                    vec![format!("fixtures/{datapackage1}"), format!("fixtures/{datapackage2}")],
                    tmp.clone()
            ).unwrap();

            insta::assert_yaml_snapshot!(format!("{name}_folder"),
                datapackage_json_to_value(&tmp.to_string_lossy().into_owned()).unwrap()
            );
            test_merged_csv_output(&tmp, format!("{name}_folder"))
        }

        {
            let temp_dir = TempDir::new().unwrap();
            let tmp = temp_dir.path().to_path_buf();

            merge_datapackage(
                    vec![format!("fixtures/{datapackage1}.zip"), format!("fixtures/{datapackage2}.zip")],
                    tmp.clone()
            ).unwrap();

            insta::assert_yaml_snapshot!(format!("{name}_zip"),
                datapackage_json_to_value(&tmp.to_string_lossy().into_owned()).unwrap()
            );
            test_merged_csv_output(&tmp, format!("{name}_zip"))
        }
    }

    #[test]
    fn test_datapackage_merge_self() {
        test_datapackage_merge("base","base_datapackage", "base_datapackage");
    }

    #[test]
    fn test_datapackage_add_resource() {
        test_datapackage_merge("add_resource","base_datapackage", "add_resource");
    }

    #[test]
    fn test_datapackage_add_different_resource() {
        test_datapackage_merge("add_different_resource","base_datapackage", "add_different_resource");
    }

    #[test]
    fn test_datapackage_add_field() {
        test_datapackage_merge("add_field","base_datapackage", "add_field");
    }

    #[test]
    fn test_conflict_types() {
        test_datapackage_merge("conflict_types","base_datapackage", "conflict_types");
    }

    #[test]
    fn test_multiple() {
        insta::assert_yaml_snapshot!(merge_datapackage_jsons(
            vec!["fixtures/base_datapackage/datapackage.json".into(),
                 "fixtures/base_datapackage/datapackage.json".into(),
                 "fixtures/add_resource/datapackage.json".into(),
                 "fixtures/add_different_resource/datapackage.json".into(),
                 "fixtures/add_field/datapackage.json".into(),
                 "fixtures/conflict_types/datapackage.json".into()]
                ).unwrap());
    }

}