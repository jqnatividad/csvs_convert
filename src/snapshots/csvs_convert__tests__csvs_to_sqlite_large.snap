---
source: src/lib.rs
expression: datapackage
---
profile: tabular-data-package
resources:
  - profile: tabular-data-resource
    name: data
    row_count: 83192
    schema:
      fields:
        - name: _link
          type: string
          format: string
          stats:
            min_len: 8
            max_len: 12
            min_str: 0.data.0
            max_str: 999.data.9
            min_number: ~
            max_number: ~
            count: 83192
            empty_count: 0
            estimate_unique: 82252
            sum: ~
            mean: ~
        - name: _link_daily_16
          type: integer
          format: integer
          stats:
            min_len: 1
            max_len: 4
            min_str: "0"
            max_str: "999"
            min_number: 0.0
            max_number: 4998.0
            count: 83192
            empty_count: 0
            estimate_unique: 5102
            sum: 208553143.0
            mean: 2506.8894004231174
        - name: dt
          type: integer
          format: integer
          stats:
            min_len: 10
            max_len: 10
            min_str: "1489395600"
            max_str: "1490788800"
            min_number: 1489395600.0
            max_number: 1490788800.0
            count: 83192
            empty_count: 0
            estimate_unique: 333
            sum: 123963937516800.0
            mean: 1490094450.3894606
        - name: rain
          type: number
          format: number
          stats:
            min_len: 1
            max_len: 5
            min_str: "0.2"
            max_str: "95.19"
            min_number: 0.2
            max_number: 95.19
            count: 50409
            empty_count: 32783
            estimate_unique: 2186
            sum: 206982.71999999954
            mean: 4.106066773790386
        - name: pressure
          type: number
          format: number
          stats:
            min_len: 3
            max_len: 7
            min_str: "1000"
            max_str: "999.99"
            min_number: 545.07
            max_number: 1050.13
            count: 83192
            empty_count: 0
            estimate_unique: 12531
            sum: 81331797.01000002
            mean: 977.6396409510533
        - name: humidity
          type: integer
          format: integer
          stats:
            min_len: 1
            max_len: 3
            min_str: "0"
            max_str: "99"
            min_number: 0.0
            max_number: 100.0
            count: 83192
            empty_count: 0
            estimate_unique: 93
            sum: 1152691.0
            mean: 13.855791422252139
        - name: uvi
          type: number
          format: number
          stats:
            min_len: 1
            max_len: 5
            min_str: "0.43"
            max_str: "9.99"
            min_number: 0.43
            max_number: 18.36
            count: 83192
            empty_count: 0
            estimate_unique: 1593
            sum: 456544.6999999962
            mean: 5.487843783056018
        - name: speed
          type: number
          format: number
          stats:
            min_len: 1
            max_len: 5
            min_str: "0.04"
            max_str: "9.99"
            min_number: 0.04
            max_number: 19.05
            count: 83192
            empty_count: 0
            estimate_unique: 1349
            sum: 337691.79000000074
            mean: 4.059185859217242
        - name: deg
          type: integer
          format: integer
          stats:
            min_len: 1
            max_len: 3
            min_str: "0"
            max_str: "99"
            min_number: 0.0
            max_number: 360.0
            count: 83192
            empty_count: 0
            estimate_unique: 362
            sum: 14358770.0
            mean: 172.5979661505914
        - name: clouds
          type: integer
          format: integer
          stats:
            min_len: 1
            max_len: 3
            min_str: "0"
            max_str: "99"
            min_number: 0.0
            max_number: 100.0
            count: 83192
            empty_count: 0
            estimate_unique: 102
            sum: 3061459.0
            mean: 36.7999206654486
        - name: temp_day
          type: number
          format: number
          stats:
            min_len: 3
            max_len: 6
            min_str: "246.7"
            max_str: "316.53"
            min_number: 246.7
            max_number: 316.53
            count: 83192
            empty_count: 0
            estimate_unique: 4750
            sum: 23834088.99999997
            mean: 286.4949634580245
        - name: temp_min
          type: number
          format: number
          stats:
            min_len: 3
            max_len: 6
            min_str: "239.15"
            max_str: "303"
            min_number: 239.15
            max_number: 303.0
            count: 83192
            empty_count: 0
            estimate_unique: 4667
            sum: 23280137.579999942
            mean: 279.83625324550366
        - name: temp_max
          type: number
          format: number
          stats:
            min_len: 3
            max_len: 6
            min_str: "246.7"
            max_str: "316.53"
            min_number: 246.7
            max_number: 316.53
            count: 83192
            empty_count: 0
            estimate_unique: 4752
            sum: 23873359.639999926
            mean: 286.96701173189643
        - name: temp_night
          type: number
          format: number
          stats:
            min_len: 3
            max_len: 6
            min_str: "240.31"
            max_str: "312.75"
            min_number: 240.31
            max_number: 312.75
            count: 83192
            empty_count: 0
            estimate_unique: 4550
            sum: 23409728.579999797
            mean: 281.3939871622247
        - name: temp_eve
          type: number
          format: number
          stats:
            min_len: 3
            max_len: 6
            min_str: "246.7"
            max_str: "313.01"
            min_number: 246.7
            max_number: 313.01
            count: 83192
            empty_count: 0
            estimate_unique: 4663
            sum: 23692752.860000007
            mean: 284.7960484181172
        - name: temp_morn
          type: number
          format: number
          stats:
            min_len: 3
            max_len: 6
            min_str: "239.15"
            max_str: "314.62"
            min_number: 239.15
            max_number: 314.62
            count: 83192
            empty_count: 0
            estimate_unique: 4684
            sum: 23355239.389999975
            mean: 280.73900603423374
        - name: snow
          type: number
          format: number
          stats:
            min_len: 1
            max_len: 5
            min_str: "0"
            max_str: "9.91"
            min_number: 0.0
            max_number: 46.72
            count: 13273
            empty_count: 69919
            estimate_unique: 686
            sum: 12979.64999999996
            mean: 0.9778987418066721
    path: fixtures/large/csv/data.csv
    dialect:
      delimiter: ","
      quoteChar: "\""
  - profile: tabular-data-resource
    name: daily_16
    row_count: 4999
    schema:
      fields:
        - name: _link
          type: integer
          format: integer
          stats:
            min_len: 1
            max_len: 4
            min_str: "0"
            max_str: "999"
            min_number: 0.0
            max_number: 4998.0
            count: 4999
            empty_count: 0
            estimate_unique: 5102
            sum: 12492501.0
            mean: 2499.0
        - name: time
          type: integer
          format: integer
          stats:
            min_len: 10
            max_len: 10
            min_str: "1489458708"
            max_str: "1489459697"
            min_number: 1489458708.0
            max_number: 1489459697.0
            count: 4999
            empty_count: 0
            estimate_unique: 969
            sum: 7445807130574.0
            mean: 1489459317.9783957
        - name: city_id
          type: integer
          format: integer
          stats:
            min_len: 5
            max_len: 8
            min_str: "1001642"
            max_str: "998210"
            min_number: 34915.0
            max_number: 10722858.0
            count: 4999
            empty_count: 0
            estimate_unique: 4972
            sum: 17587664942.0
            mean: 3518236.6357271452
        - name: city_name
          type: string
          format: string
          stats:
            min_len: 1
            max_len: 55
            min_str: "-"
            max_str: 괴내
            min_number: ~
            max_number: ~
            count: 4999
            empty_count: 0
            estimate_unique: 4953
            sum: ~
            mean: ~
        - name: city_country
          type: string
          format: string
          stats:
            min_len: 2
            max_len: 2
            min_str: AF
            max_str: ZW
            min_number: ~
            max_number: ~
            count: 4995
            empty_count: 4
            estimate_unique: 169
            sum: ~
            mean: ~
        - name: city_coord_lon
          type: number
          format: number
          stats:
            min_len: 1
            max_len: 11
            min_str: "-0.01667"
            max_str: "99.966667"
            min_number: -170.649994
            max_number: 179.383331
            count: 4999
            empty_count: 0
            estimate_unique: 4677
            sum: 80462.53737899997
            mean: 16.095726621124218
        - name: city_coord_lat
          type: number
          format: number
          stats:
            min_len: 1
            max_len: 10
            min_str: "-0.08333"
            max_str: "9.96667"
            min_number: -54.281109
            max_number: 70.663361
            count: 4999
            empty_count: 0
            estimate_unique: 4377
            sum: 175013.376443
            mean: 35.00967722404481
    path: fixtures/large/csv/daily_16.csv
    dialect:
      delimiter: ","
      quoteChar: "\""
  - profile: tabular-data-resource
    name: data_weather
    row_count: 83192
    schema:
      fields:
        - name: _link
          type: string
          format: string
          stats:
            min_len: 18
            max_len: 22
            min_str: 0.data.0.weather.0
            max_str: 999.data.9.weather.0
            min_number: ~
            max_number: ~
            count: 83192
            empty_count: 0
            estimate_unique: 83396
            sum: ~
            mean: ~
        - name: _link_data
          type: string
          format: string
          stats:
            min_len: 8
            max_len: 12
            min_str: 0.data.0
            max_str: 999.data.9
            min_number: ~
            max_number: ~
            count: 83192
            empty_count: 0
            estimate_unique: 82252
            sum: ~
            mean: ~
        - name: _link_daily_16
          type: integer
          format: integer
          stats:
            min_len: 1
            max_len: 4
            min_str: "0"
            max_str: "999"
            min_number: 0.0
            max_number: 4998.0
            count: 83192
            empty_count: 0
            estimate_unique: 5102
            sum: 208553143.0
            mean: 2506.8894004231174
        - name: id
          type: integer
          format: integer
          stats:
            min_len: 3
            max_len: 3
            min_str: "500"
            max_str: "804"
            min_number: 500.0
            max_number: 804.0
            count: 83192
            empty_count: 0
            estimate_unique: 12
            sum: 49929768.0
            mean: 600.1751129916338
        - name: main
          type: string
          format: string
          stats:
            min_len: 4
            max_len: 6
            min_str: Clear
            max_str: Snow
            min_number: ~
            max_number: ~
            count: 83192
            empty_count: 0
            estimate_unique: 4
            sum: ~
            mean: ~
        - name: description
          type: string
          format: string
          stats:
            min_len: 4
            max_len: 20
            min_str: broken clouds
            max_str: very heavy rain
            min_number: ~
            max_number: ~
            count: 83192
            empty_count: 0
            estimate_unique: 12
            sum: ~
            mean: ~
        - name: icon
          type: string
          format: string
          stats:
            min_len: 3
            max_len: 5
            min_str: 01d
            max_str: 13d
            min_number: ~
            max_number: ~
            count: 83192
            empty_count: 0
            estimate_unique: 15
            sum: ~
            mean: ~
    path: fixtures/large/csv/data_weather.csv
    dialect:
      delimiter: ","
      quoteChar: "\""

